src/Callable.cpp

```diff
@@ -0,0 +1,178 @@
+                   << "Expected exactly " << (required_arg_count - hidden_args) << " arguments, "
+                   << "but saw " << (argc - hidden_args) << ".";
+    } else {
+        const Argument &a = contents->jit_cache.arguments.at(bad_idx);
+        const char *kind = a.is_scalar() ? "scalar" : "buffer";
+        // Note that we don't report the "actual type" here, just the expected type...
+        // saving the actual type leads to more code bloat than we can justify
+        // for this. (Consider adding as a debug-only enhancement?)
+        user_error << "Error " << verb << " '" << contents->name << "':\n"
+                   << "Argument " << (bad_idx - hidden_args + 1)
+                   << " of " << (required_arg_count - hidden_args) << " ('" << a.name << "') was expected to be a "
+                   << kind << " of type '" << a.type << "' and dimension " << (int)a.dimensions << ".\n";
+    }
+}
+
+void Callable::check_qcci(size_t argc, const QuickCallCheckInfo *actual_qcci) const {
+    const size_t required_arg_count = contents->quick_call_check_info.size();
+    if (argc == required_arg_count) {
+        const QuickCallCheckInfo *expected_qcci = contents->quick_call_check_info.data();
+        for (size_t i = 0; i < argc; i++) {
+            if (actual_qcci[i] != expected_qcci[i]) {
+                do_check_fail(i, argc, "calling");
+            }
+        }
+    } else {
+        do_check_fail(-1, argc, "calling");
+    }
+}
+
+void Callable::check_fcci(size_t argc, const FullCallCheckInfo *actual_fcci) const {
+    // Lazily create full_call_check_info upon the first call to make_std_function().
+    if (contents->full_call_check_info.empty()) {
+        contents->full_call_check_info.reserve(contents->jit_cache.arguments.size());
+        for (const Argument &a : contents->jit_cache.arguments) {
+            const auto fcci = a.is_scalar() ? Callable::make_scalar_fcci(a.type) : Callable::make_buffer_fcci(a.type, a.dimensions);
+            contents->full_call_check_info.push_back(fcci);
+        }
+    }
+
+    const size_t required_arg_count = contents->full_call_check_info.size();
+    if (argc == required_arg_count) {
+        const FullCallCheckInfo *expected_fcci = contents->full_call_check_info.data();
+        for (size_t i = 0; i < argc; i++) {
+            if (!Callable::is_compatible_fcci(actual_fcci[i], expected_fcci[i])) {
+                do_check_fail(i, argc, "defining");
+            }
+        }
+    } else {
+        do_check_fail(-1, argc, "defining");
+    }
+}
+
+// Entry point used from the std::function<> variant; we can skip the check_qcci() stuff
+// since we verified the signature when we created the std::function, so incorrect types or counts
+// should be impossible.
+/*static*/ int Callable::call_argv_fast(size_t argc, const void *const *argv) const {
+    // Callable should enforce these, so we can use assert() instead of internal_assert()
```

Comment:
I'm confused by this. Don't we use internal_assert for checking invariants? Or is the claim that these are already checked elsewhere and this is effectively just documentation?

Reply:
The latter -- I'll add a comment to that effect. The idea here is to avoid the cost of internal_assert() in this case since it's theoretically redundant here.